lib/api/src/modules/stories.ts

```diff
@@ -368,10 +375,36 @@ export const init: ModuleFn = ({
       }
     });
 
-    fullAPI.on(STORY_ARGS_UPDATED, (id: StoryId, args: Args) => {
-      const { storiesHash } = store.getState();
-      (storiesHash[id] as Story).args = args;
-      store.setState({ storiesHash });
+    fullAPI.on(STORY_ARGS_UPDATED, function handleStoryArgsUpdated({
+      storyId,
+      args,
+    }: {
+      storyId: StoryId;
+      args: Args;
+    }) {
+      // the event originates from an iframe, event.source is the iframe's location origin + pathname
+      const { source }: { source: string } = this;
+      const [sourceType, sourceLocation] = getSourceType(source);
+
+      switch (sourceType) {
+        case 'local': {
+          const { storiesHash } = store.getState();
+          (storiesHash[storyId] as Story).args = args;
+          store.setState({ storiesHash });
+          break;
+        }
+        case 'external': {
+          const refs = fullAPI.getRefs();
+          const ref = fullAPI.findRef(sourceLocation);
+          (ref.stories[storyId] as Story).args = args;
+          store.setState({ refs: { ...refs, [ref.id]: ref } });
+          break;
+        }
+        default: {
+          logger.warn('received a STORY_ARGS_UPDATED frame that was not configured as a ref');
+          break;
+        }
```

Comment:
I updated the code here:
https://github.com/storybookjs/storybook/pull/11392

So there's much less boilerplate and duplicated code

Reply:
Yeah I know we can update this afterwards.